packages/dds/tree/src/util/breakable.ts

```diff
@@ -124,6 +143,71 @@ export function throwIfBroken<
 		this.breaker.use();
 		return target.call(this, ...args);
 	}
-
+	markBreaker(replacementMethod);
+	nameFunctionFrom(replacementMethod, target);
 	return replacementMethod as Target;
 }
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type PossiblyNamedFunction = Function & { displayName?: undefined | string };
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+function nameFunctionFrom(toName: Function, nameFrom: Function): void {
+	(toName as PossiblyNamedFunction).displayName =
+		(nameFrom as PossiblyNamedFunction).displayName ?? nameFrom.name;
+}
+
+const isBreakerSymbol: unique symbol = Symbol("isBreaker");
+
+// Accepting any function like value is desired and safe here as this does not call the provided function.
+// eslint-disable-next-line @typescript-eslint/ban-types
+function markBreaker(f: Function): void {
+	(f as unknown as Record<typeof isBreakerSymbol, true>)[isBreakerSymbol] = true;
+}
+
+// Accepting any function like value is desired and safe here as this does not call the provided function.
+// eslint-disable-next-line @typescript-eslint/ban-types
+function isBreaker(f: Function): boolean {
+	return isBreakerSymbol in (f as unknown as Record<typeof isBreakerSymbol, true>);
+}
+
+/**
+ * Decorator for classes which should break when their methods throw.
+ * @remarks
+ * Applies {@link breakingMethod} to all methods declared directly class or its base classes.
+ * Does not include those on derived classes.
+ * Does not include getters or setters, or value properties.
+ * Methods already marked as {@link breakingMethod} or {@link throwIfBroken} are unaffected.
+ */
+export function breakingClass<Target extends abstract new (...args: any[]) => WithBreakable>(
+	target: Target,
+	context: ClassDecoratorContext<Target>,
+): Target {
+	abstract class DecoratedBreakable extends target {}
+
+	// Keep track of what keys we have seen,
+	// since we visit most derived properties first and need to avoid wrapping base properties overriding more derived ones.
+	const overriddenKeys: Set<string | symbol> = new Set();
+
+	let prototype: object | null = target.prototype;
+	while (prototype !== null) {
+		for (const key of Reflect.ownKeys(prototype)) {
+			if (!overriddenKeys.has(key)) {
+				overriddenKeys.add(key);
+				const descriptor = Reflect.getOwnPropertyDescriptor(prototype, key);
+				if (descriptor !== undefined) {
+					// Method
+					if (typeof descriptor.value === "function") {
+						if (!isBreaker(descriptor.value)) {
+							descriptor.value = breakingMethod(descriptor.value);
```

Comment:
Does this assignment affect the original object? If so, why do we want that since we are returning a different class?

Reply:
No. See first paragraph in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor

Reply:
```suggestion
							// This does not effect the original class, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor
							descriptor.value = breakingMethod(descriptor.value);
```